File: Second_semestr/Homework_2/HashTable/HashTable/HashTable.cs
using System;
using ListNamespace;

namespace HashTableNamespace
{
    /// <summary>
    /// a class that represents a hash table
    /// </summary>
    public class HashTable : IHashTable
    {
        /// <summary>
        /// class constructor
        /// </summary>
        public HashTable()
        {
            hashTable = new List[101];
            for (int i = 0; i < 101; ++i)
            {
                hashTable[i] = new List();
            }
        }

        public void Add(string word)
        {
            int i = this.hashFunction(word);
            if (!this.Contains(word))
            {
                this.hashTable[i].Add(word);
            }
            else
            {
                Console.WriteLine("this item already exists in the hash table");
            }
        }

        public void Remove(string word)
        {
            int i = hashFunction(word);
            hashTable[i].Remove(word);
        }

        public bool Contains(string word)
        {
            int i = hashFunction(word);
            return hashTable[i].Contains(word);
        }

        public void Print()
        {
            Console.WriteLine("Hash table:");
            int size = hashTable.GetLength(0);
            for (int i = 0; i < size; ++i)
            {
                Console.WriteLine("List number " + i + ":");
                hashTable[i].Print();
                Console.WriteLine();
            }
        }

        private List[] hashTable;

        private int hashFunction(string word)
        {
            int result = 0;
            int length = word.Length;
            for (int i = 0; i < length; ++i)
            {
                result += word[i];
            }

            return result % hashTable.GetLength(0);
        }
    }
}

File: Fourth_semester/test/test/HashTable.fs
module HashTable

type HashTable(hashFunc) =
    let hashTable = Array.init 101 (fun index -> List.empty)
    
    member this.Add x =
        let i = (hashFunc x) % 101
        if this.Contains(x) then
            printfn "%A" "this item already exists in the hash table"
        else
            hashTable.[i] <- List.Cons (x, hashTable.[i])

    member this.Contains x =
        let i = hashFunc x
        List.contains x hashTable.[i]
    
    member this.Remove x =
        let i = hashFunc x
        hashTable.[i] <- List.filter (fun elem -> x <> elem) hashTable.[i]

    member this.Print () =
        printfn "%A" "Hash table:"
        Seq.iteri (fun i elem -> printfn "%A" ("List number " + i.ToString() + ":")
                                 List.iter (fun x -> printf "%s" x) elem) hashTable